File: crates/riptide-api/src/handlers/spider.rs
//! Spider handlers using SpiderFacade
//!
//! This module provides HTTP handlers for deep crawling operations using
//! the riptide-facade SpiderFacade for simplified spider engine access.
#![allow(dead_code)]
use crate::context::ApplicationContext;
use crate::errors::ApiError;
use crate::models::*;
use axum::{
    extract::{Query, State},
    response::IntoResponse,
    Json,
};
use riptide_types::ResultMode;
use serde::Deserialize;

// Unused imports removed - parse_seed_urls and MetricsRecorder not used yet

/// Default max content size in bytes (1MB)
#[allow(dead_code)]
const DEFAULT_MAX_CONTENT_BYTES: usize = 1_048_576;

/// Query parameters for spider crawl endpoint
#[derive(Debug, Deserialize)]
pub struct SpiderCrawlQuery {
    /// Result mode: stats (default), urls, pages, stream, or store
    #[serde(default)]
    pub result_mode: ResultMode,

    /// Include specific fields (comma-separated, e.g., "title,links,markdown")
    pub include: Option<String>,

    /// Exclude specific fields (comma-separated, e.g., "content")
    pub exclude: Option<String>,

    /// Maximum content size in bytes per page (default: 1MB)
    pub max_content_bytes: Option<usize>,
}

/// Spider crawl endpoint for deep crawling operations.
///
/// This endpoint uses the SpiderFacade to perform deep crawling with:
/// - Simplified facade API over the Spider engine
/// - Frontier-based URL queue management
/// - Multiple crawling strategies (BFS, DFS, Best-First)
/// - Adaptive stopping based on content analysis
/// - Budget controls and rate limiting
/// - Session persistence for authenticated crawling
///
/// Query Parameters:
/// - result_mode: "stats" (default) or "urls"
///   - stats: Returns statistics only (backward compatible)
///   - urls: Returns statistics plus discovered URLs list
#[tracing::instrument(
    name = "spider_crawl",
    skip_all,
    fields(
        http.method = "POST",
        http.route = "/spider/crawl",
        seed_count = body.seed_urls.len(),
        max_depth = body.max_depth,
        max_pages = body.max_pages,
        otel.status_code
    )
)]
pub async fn spider_crawl(
    State(_state): State<ApplicationContext>,
    Query(_query): Query<SpiderCrawlQuery>,
    Json(body): Json<SpiderCrawlBody>,
) -> Result<impl IntoResponse, ApiError> {
    // Acquire resources via ResourceFacade (Phase 5 - Handler Integration)
    let tenant_id = "spider-crawl"; // TODO: Extract from request context in Phase 6
    let _resource_slot = acquire_spider_resources(&_state, tenant_id).await?;

    // Get spider facade
    let spider_facade = _state
        .spider_facade
        .as_ref()
        .ok_or_else(|| ApiError::ConfigError {
            message: "SpiderFacade not initialized".to_string(),
        })?;

    // Parse seed URLs
    let seed_urls: Vec<url::Url> = body
        .seed_urls
        .iter()
        .map(|s| url::Url::parse(s))
        .collect::<Result<Vec<_>, _>>()
        .map_err(|e| ApiError::InvalidUrl {
            url: "".to_string(),
            message: format!("Invalid seed URL: {}", e),
        })?;

    // Execute crawl via facade
    let summary = spider_facade
        .crawl(seed_urls)
        .await
        .map_err(|e| ApiError::InternalError {
            message: format!("Spider crawl failed: {}", e),
        })?;

    Ok(Json(summary))
}

/// Get spider status and metrics
pub async fn spider_status(
    State(_state): State<ApplicationContext>,
    Json(_body): Json<SpiderStatusRequest>,
) -> Result<impl IntoResponse, ApiError> {
    // Get spider facade
    let spider_facade = _state
        .spider_facade
        .as_ref()
        .ok_or_else(|| ApiError::ConfigError {
            message: "SpiderFacade not initialized".to_string(),
        })?;

    // Get status from facade
    let (state, performance) =
        spider_facade
            .get_status()
            .await
            .map_err(|e| ApiError::InternalError {
                message: format!("Failed to get spider status: {}", e),
            })?;

    // Build response
    let response = SpiderStatusResponse {
        state,
        performance,
        frontier_stats: None,
        adaptive_stop_stats: None,
    };

    Ok(Json(response))
}

/// Spider control endpoint for start/stop/reset operations
pub async fn spider_control(
    State(_state): State<ApplicationContext>,
    Json(_body): Json<SpiderControlRequest>,
) -> Result<impl IntoResponse, ApiError> {
    // Get spider facade
    let spider_facade = _state
        .spider_facade
        .as_ref()
        .ok_or_else(|| ApiError::ConfigError {
            message: "SpiderFacade not initialized".to_string(),
        })?;

    // Execute control action via facade
    let message =
        spider_facade
            .control(&_body.action)
            .await
            .map_err(|e| ApiError::ValidationError {
                message: e.to_string(),
            })?;

    Ok(Json(serde_json::json!({
        "success": true,
        "message": message
    })))
}

/// Acquire resources for spider operations via ResourceFacade
///
/// Coordinates rate limiting, memory pressure, and pool capacity through
/// the ResourceFacade layer (Phase 5 integration).
async fn acquire_spider_resources(
    state: &ApplicationContext,
    tenant_id: &str,
) -> Result<crate::adapters::ResourceSlot, ApiError> {
    use riptide_facade::facades::ResourceResult as FacadeResult;

    // Acquire WASM slot through facade (handles all resource coordination)
    match state.resource_facade.acquire_wasm_slot(tenant_id).await {
        Ok(FacadeResult::Success(slot)) => Ok(slot),
        Ok(FacadeResult::RateLimited { retry_after }) => Err(ApiError::RateLimitExceeded {
            retry_after: retry_after.as_secs(),
        }),
        Ok(FacadeResult::MemoryPressure) => Err(ApiError::InternalError {
            message: "System under memory pressure".to_string(),
        }),
        Ok(FacadeResult::ResourceExhausted) => Err(ApiError::InternalError {
            message: "Spider resources exhausted".to_string(),
        }),
        Ok(FacadeResult::Timeout) => Err(ApiError::TimeoutError {
            operation: "Resource acquisition".to_string(),
            message: "Timeout acquiring spider resources".to_string(),
        }),
        Err(e) => Err(ApiError::InternalError {
            message: format!("Resource facade error: {}", e),
        }),
    }
}

File: sdk/python/riptide_sdk/endpoints/spider.py
"""
Spider crawling API endpoint implementation

Provides deep crawling operations using the Spider engine with:
- Frontier-based URL queue management
- Multiple crawling strategies (BFS, DFS, Best-First)
- Adaptive stopping based on content analysis
- Budget controls and rate limiting
- Session persistence for authenticated crawling
"""

from typing import List, Optional, Dict, Any, Literal
import httpx

from ..models import (
    SpiderConfig,
    SpiderResult,
    SpiderStatus,
    SpiderControlResponse,
    ResultMode,
)
from ..exceptions import APIError, ValidationError, ConfigError


class SpiderAPI:
    """
    API for deep crawling operations using Spider engine

    The Spider API provides advanced crawling capabilities beyond simple
    batch operations, including multi-page crawling with intelligent
    frontier management and adaptive stopping conditions.
    """

    def __init__(self, client: httpx.AsyncClient, base_url: str):
        """
        Initialize SpiderAPI

        Args:
            client: Async HTTP client
            base_url: Base URL for the API
        """
        self.client = client
        self.base_url = base_url

    async def crawl(
        self,
        seed_urls: List[str],
        config: Optional[SpiderConfig] = None,
        result_mode: ResultMode = ResultMode.STATS,
    ) -> SpiderResult:
        """
        Start a deep crawl from seed URLs using the Spider engine

        This endpoint performs multi-page crawling with intelligent URL frontier
        management, multiple strategies, and adaptive stopping conditions.

        Args:
            seed_urls: List of starting URLs for the crawl
            config: Optional spider configuration (defaults, depth, strategy, etc.)
            result_mode: Result mode - STATS (default) or URLS to include discovered URLs

        Returns:
            SpiderResult with crawl summary, state, and performance metrics.
            If result_mode=URLS, also includes discovered_urls list.

        Raises:
            ValidationError: If seed URLs are invalid or empty
            ConfigError: If SpiderFacade is not enabled on the server
            APIError: If the API returns an error

        Example:
            Basic crawl (stats only):
            >>> result = await client.spider.crawl(
            ...     seed_urls=["https://example.com"],
            ... )
            >>> print(f"Crawled {result.pages_crawled} pages")
            >>> print(f"Stop reason: {result.stop_reason}")

            Crawl with URLs mode (for discovery):
            >>> from riptide_sdk import ResultMode
            >>> result = await client.spider.crawl(
            ...     seed_urls=["https://example.com"],
            ...     result_mode=ResultMode.URLS,
            ... )
            >>> print(f"Discovered {len(result.discovered_urls)} URLs")
            >>> for url in result.discovered_urls[:10]:
            ...     print(f"  - {url}")

            Advanced configuration:
            >>> from riptide_sdk import SpiderConfig, ResultMode
            >>> config = SpiderConfig(
            ...     max_depth=3,
            ...     max_pages=100,
            ...     strategy="breadth_first",
            ...     concurrency=10,
            ...     delay_ms=500,
            ...     respect_robots=True,
            ... )
            >>> result = await client.spider.crawl(
            ...     seed_urls=["https://example.com"],
            ...     config=config,
            ...     result_mode=ResultMode.URLS,
            ... )
            >>> print(result.to_summary())
            >>> print(f"Discovered URLs: {len(result.discovered_urls)}")

            Discover → Extract workflow:
            >>> # Step 1: Discover URLs
            >>> discovery = await client.spider.crawl(
            ...     seed_urls=["https://example.com/blog"],
            ...     config=SpiderConfig(max_depth=2),
            ...     result_mode=ResultMode.URLS,
            ... )
            >>> # Step 2: Extract content from discovered URLs
            >>> for url in discovery.discovered_urls:
            ...     content = await client.extract.extract(url)
            ...     print(f"Extracted: {content.title}")
        """
        if not seed_urls:
            raise ValidationError("seed_urls list cannot be empty")

        if len(seed_urls) > 50:
            raise ValidationError("Maximum 50 seed URLs per crawl request")

        # Validate seed URLs
        for url in seed_urls:
            if not url.startswith(("http://", "https://")):
                raise ValidationError(f"Invalid seed URL: {url}")

        # Build request body
        body = {"seed_urls": seed_urls}
        if config:
            body.update(config.to_dict())

        # Build query parameters
        params = {}
        if result_mode != ResultMode.STATS:
            params["result_mode"] = result_mode.value

        # Make request
        try:
            response = await self.client.post(
                f"{self.base_url}/api/v1/spider/crawl",
                json=body,
                params=params,
            )
        except httpx.RequestError as e:
            raise APIError(
                message=f"Request failed: {str(e)}",
                status_code=0,
            )

        if response.status_code == 500:
            error_data = response.json() if response.text else {}
            error_msg = error_data.get("error", {}).get("message", "")
            if "SpiderFacade is not enabled" in error_msg:
                raise ConfigError(
                    "SpiderFacade is not enabled on the server. "
                    "Please enable spider functionality in server configuration."
                )
            raise APIError(
                message=error_msg or "Spider crawl failed",
                status_code=response.status_code,
                response_data=error_data,
            )

        if response.status_code != 200:
            error_data = response.json() if response.text else {}
            raise APIError(
                message=error_data.get("error", {}).get("message", "Spider crawl failed"),
                status_code=response.status_code,
                response_data=error_data,
            )

        return SpiderResult.from_dict(response.json())

    async def status(
        self,
        include_metrics: bool = False,
    ) -> SpiderStatus:
        """
        Get current spider status and metrics

        Retrieves the current state of the spider engine, including active
        crawl status, pages processed, and optionally detailed performance metrics.

        Args:
            include_metrics: Whether to include detailed performance metrics

        Returns:
            SpiderStatus with current state and optional metrics

        Raises:
            ConfigError: If SpiderFacade is not enabled
            APIError: If the API returns an error

        Example:
            Basic status check:
            >>> status = await client.spider.status()
            >>> print(f"Active: {status.state.active}")
            >>> print(f"Pages crawled: {status.state.pages_crawled}")

            With detailed metrics:
            >>> status = await client.spider.status(include_metrics=True)
            >>> if status.performance:
            ...     print(f"Pages/sec: {status.performance.pages_per_second:.2f}")
            ...     print(f"Error rate: {status.performance.error_rate:.2%}")
        """
        body = {"include_metrics": include_metrics}

        try:
            response = await self.client.post(
                f"{self.base_url}/api/v1/spider/status",
                json=body,
            )
        except httpx.RequestError as e:
            raise APIError(
                message=f"Request failed: {str(e)}",
                status_code=0,
            )

        if response.status_code == 500:
            error_data = response.json() if response.text else {}
            error_msg = error_data.get("error", {}).get("message", "")
            if "SpiderFacade is not enabled" in error_msg:
                raise ConfigError(
                    "SpiderFacade is not enabled on the server. "
                    "Please enable spider functionality in server configuration."
                )
            raise APIError(
                message=error_msg or "Failed to get spider status",
                status_code=response.status_code,
                response_data=error_data,
            )

        if response.status_code != 200:
            error_data = response.json() if response.text else {}
            raise APIError(
                message=error_data.get("error", {}).get("message", "Failed to get spider status"),
                status_code=response.status_code,
                response_data=error_data,
            )

        return SpiderStatus.from_dict(response.json())

    async def control(
        self,
        action: Literal["stop", "reset"],
    ) -> SpiderControlResponse:
        """
        Control spider crawling operations

        Allows controlling the spider engine state through various actions:
        - stop: Stop the current crawl operation
        - reset: Reset the spider state and clear frontier

        Args:
            action: Control action to perform ("stop" or "reset")

        Returns:
            SpiderControlResponse with action result status

        Raises:
            ValidationError: If action is invalid
            ConfigError: If SpiderFacade is not enabled
            APIError: If the API returns an error

        Example:
            Stop current crawl:
            >>> response = await client.spider.control(action="stop")
            >>> print(response.status)  # "stopped"

            Reset spider state:
            >>> response = await client.spider.control(action="reset")
            >>> print(response.status)  # "reset"

            Error handling:
            >>> try:
            ...     await client.spider.control(action="pause")
            ... except ValidationError as e:
            ...     print(f"Invalid action: {e}")
        """
        if action not in ("stop", "reset"):
            raise ValidationError(
                f"Invalid action '{action}'. Must be 'stop' or 'reset'"
            )

        body = {"action": action}

        try:
            response = await self.client.post(
                f"{self.base_url}/api/v1/spider/control",
                json=body,
            )
        except httpx.RequestError as e:
            raise APIError(
                message=f"Request failed: {str(e)}",
                status_code=0,
            )

        if response.status_code == 500:
            error_data = response.json() if response.text else {}
            error_msg = error_data.get("error", {}).get("message", "")
            if "SpiderFacade is not enabled" in error_msg:
                raise ConfigError(
                    "SpiderFacade is not enabled on the server. "
                    "Please enable spider functionality in server configuration."
                )
            raise APIError(
                message=error_msg or f"Spider {action} failed",
                status_code=response.status_code,
                response_data=error_data,
            )

        if response.status_code != 200:
            error_data = response.json() if response.text else {}
            raise APIError(
                message=error_data.get("error", {}).get("message", f"Spider {action} failed"),
                status_code=response.status_code,
                response_data=error_data,
            )

        result = response.json()
        return SpiderControlResponse(status=result.get("status", action))

    async def crawl_with_status_polling(
        self,
        seed_urls: List[str],
        config: Optional[SpiderConfig] = None,
        poll_interval: float = 2.0,
        callback: Optional[callable] = None,
    ) -> SpiderResult:
        """
        Start a crawl and poll status until completion

        This is a convenience method that starts a crawl and periodically
        checks the status until the crawl completes. Useful for long-running
        crawls where you want progress updates.

        Args:
            seed_urls: List of starting URLs for the crawl
            config: Optional spider configuration
            poll_interval: Seconds between status checks (default: 2.0)
            callback: Optional callback function called with each status update

        Returns:
            SpiderResult with final crawl summary

        Example:
            With progress callback:
            >>> def on_progress(status):
            ...     print(f"Crawled: {status.state.pages_crawled}")
            >>>
            >>> result = await client.spider.crawl_with_status_polling(
            ...     seed_urls=["https://example.com"],
            ...     config=SpiderConfig(max_pages=100),
            ...     poll_interval=1.0,
            ...     callback=on_progress,
            ... )
        """
        import asyncio

        # Start the crawl (non-blocking on server side)
        result = await self.crawl(seed_urls, config)

        # If callback provided, poll for status updates
        if callback:
            while True:
                status = await self.status(include_metrics=True)
                callback(status)

                if not status.state.active:
                    break

                await asyncio.sleep(poll_interval)

        return result
